README.txt
repoze.accelerator
==================

An caching accelerator for WSGI applications.  Intercepts responses
from downstream WSGI applications and caches responses based on
Cache-Control (and other) response header values.  Serves up cached
responses to clients thereafter based on the policy specified by the
headers.

Configuration via Python
------------------------

Configure the middleware via Python::

  import logging
  from repoze.accelerator.policy import AcceleratorPolicy
  from repoze.accelerator.storage import MemoryStorage
  from repoze.accelerator.middleware import AcceleratorMiddleware

  logger = logger.getLogger('repoze.accelerator')

  storage = MemoryStorage(logger)
  policy = AcceleratorPolicy(logger, storage)
  middleware = AcceleratorMiddleware(app, policy, logger)

Configuration via Paste Deploy
------------------------------

Configure the paste.ini file like so::

  [pipeline:main]
  pipeline = egg:Paste#cgitb
             egg:repoze.accelerator#accelerator
             myapp


The Default Policy
------------------

The default policy is the AcceleratorPolicy.  This policy has the
following features:

- Allow configuration of "vary" policies for both request headers
  and request environment values.

- Allow specification of "allowed methods"; requests which don't use
  one of these methods won't be served from cache.

- Allow specification of "honor_shift_reload": if Pragma: no-cache or
  Cache-Control: no-cache exists in the request, and this value is true,
  the response will not be served from cache even if it otherwise might
  have been.

- Allow specification of "store_https_responses".  If this is true,
  we will store https responses and some information provided by
  requests emitted via HTTPS.

When deciding whether we can fetch from our storage or not:

- If we honor shift-reload, and the request has a Pragma: no-cache
  or Cache-Control: no-cache associated with it, don't try to
  retrieve it from storage.

- If the request method doesn't match one of our allowed_methods,
  don't try to retrieve it from storage.

- If the request has a Range, header, don't try to retrieve it
  from storage.

- If the request is conditional (e.g. If-Modified-Since,
  If-None-Match), don't try to retrieve it from storage.

- Otherwise, attempt to retrieve it from storage.

When deciding whether a value returned from storage should be
checked for freshness:

- If one representation stored matches the request headers and
  environment, check it for freshness.

- Otherwise, abort.

To decide whether an item is fresh or not:

- If the entry in the cache is stale, don't serve from cache.
  Staleness is defined as having a CC: max-age < (now -
  entitydate) or an expires header whereby (expires - entitydate)
  < (now - entitydate).  Entities which don't have a Date header
  are also considered stale.

When deciding whether we can store response data in our storage:

- If the request method doesn't match one of our allowed_methods,
  don't store.

- If the response status is not 200 (OK) or 203 (Non-Authoritative
  Information), don't store.

- If the response has a Cache-Control header or a Pragma header,
  and either has 'no-cache' in its value, don't store.

- If the response has a Cache-Control header, and it has a max-age
  of 0 or a max-age we don't understand, don't store.

- If the request is an https request, and "store_https_responses" is false,
  don't store.

- If the response does not have a Date header, assume the date is now.

When storing data to storage:

- Store the status, the end-to-end headers in the response, and
  information about request header and environment variance.

Mea Culpa
---------

There is not nearly enough documentation here to be considered
canonical.  Apologies.  Please read the source for more info for now.

Reporting Bugs / Development Versions
-------------------------------------

Visit http://github.com/repoze/repoze.accelerator/issues to report bugs.
Visit http://github.com/repoze to download development or tagged versions.


CHANGES.txt
0.1
---

Initial release.



repoze/accelerator/tests.py
import unittest

_MARKER = object()

class TestMemoryStorage(unittest.TestCase):
    def _getTargetClass(self):
        from repoze.accelerator.storage import MemoryStorage
        return MemoryStorage

    def _makeOne(self, lock):
        klass = self._getTargetClass()
        return klass(lock)

    def test_class_conforms_to_IStorage(self):
        from zope.interface.verify import verifyClass
        from repoze.accelerator.interfaces import IStorage
        verifyClass(IStorage, self._getTargetClass())

    def test_instance_conforms_to_IStorage(self):
        from zope.interface.verify import verifyObject
        from repoze.accelerator.interfaces import IStorage
        verifyObject(IStorage, self._makeOne(DummyLock()))

    def test_factory_provides_IStorageFactory(self):
        from zope.interface.verify import verifyObject
        from repoze.accelerator.interfaces import IStorageFactory
        from repoze.accelerator.storage import make_memory_storage
        verifyObject(IStorageFactory, make_memory_storage)

    def test_store_nonexistent(self):
        lock = DummyLock()
        storage = self._makeOne(lock)
        headers = [('Header1', 'value1')]
        handler = storage.store('url', 'status', headers, [], [])
        self.failIf(handler is None)
        chunks = ['chunk1', 'chunk2']
        for chunk in ('chunk1', 'chunk2'):
            handler.write(chunk)
        handler.close()
        self.assertEqual(storage.data['url'][(), ()],
                         ('status', headers, chunks))
        self.assertEqual(lock.acquired, 1)
        self.assertEqual(lock.released, 1)

    def test_store_existing(self):
        lock = DummyLock()
        storage = self._makeOne(lock)
        storage.data['url'] = {}
        storage.data['url'][(), ()] = ('otherstatus', (), ())
        headers = [('Header1', 'value1')]
        handler = storage.store('url', 'status', headers, [], [])
        self.failIf(handler is None)
        chunks = ['chunk1', 'chunk2']
        for chunk in ('chunk1', 'chunk2'):
            handler.write(chunk)
        handler.close()
        self.assertEqual(storage.data['url'][(), ()],
                         ('status', headers, chunks))
        self.assertEqual(lock.acquired, 1)
        self.assertEqual(lock.released, 1)

    def test_fetch_nonexistent(self):
        lock = DummyLock()
        storage = self._makeOne(lock)
        self.assertEqual(storage.fetch('url'), None)

    def test_fetch_existing(self):
        lock = DummyLock()
        storage = self._makeOne(lock)
        storage.data['url'] = {
            (1, 2):(200, [], []),
            (3, 4):(203, [], [])
            }
        result = storage.fetch('url')
        result.sort()
        self.assertEqual(len(result), 2)
        self.assertEqual(result[0], (200, [], [], 1, 2))
        self.assertEqual(result[1], (203, [], [], 3, 4))


class TestAcceleratorPolicy(unittest.TestCase):
    def _getTargetClass(self):
        from repoze.accelerator.policy import AcceleratorPolicy
        return AcceleratorPolicy

    def _makeOne(self, storage):
        klass = self._getTargetClass()
        return klass(storage)

    def _makeEnviron(self):
        return {
            'wsgi.url_scheme':'http',
            'SERVER_NAME':'example.com',
            'SERVER_PORT':'80',
            'REQUEST_METHOD': 'GET',
            }

    def test_class_conforms_to_IPolicy(self):
        from zope.interface.verify import verifyClass
        from repoze.accelerator.interfaces import IPolicy
        verifyClass(IPolicy, self._getTargetClass())

    def test_instance_conforms_to_IPolicy(self):
        from zope.interface.verify import verifyObject
        from repoze.accelerator.interfaces import IPolicy
        verifyObject(IPolicy, self._makeOne(DummyStorage()))

    def test_factory_provides_IPolicyFactory(self):
        from zope.interface.verify import verifyObject
        from repoze.accelerator.interfaces import IPolicyFactory
        from repoze.accelerator.policy import make_accelerator_policy
        verifyObject(IPolicyFactory, make_accelerator_policy)

    def test_store_not_cacheable_post_request_method(self):
        storage = DummyStorage()
        policy = self._makeOne(storage)
        environ = self._makeEnviron()
        environ['REQUEST_METHOD'] = 'POST'
        result = policy.store('200 OK', [], environ)
        self.assertEqual(result, None)

    def test_store_not_cacheable_pragma_no_cache(self):
        storage = DummyStorage()
        policy = self._makeOne(storage)
        environ = self._makeEnviron()
        headers = [ ('pragma', 'no-cache') ]
        result = policy.store('200 OK', headers, environ)
        self.assertEqual(result, None)

    def test_store_not_cacheable_cachecontrol_no_cache(self):
        storage = DummyStorage()
        policy = self._makeOne(storage)
        environ = self._makeEnviron()
        headers = [ ('cache-control', 'no-cache') ]
        result = policy.store('200 OK', headers, environ)
        self.assertEqual(result, None)

    def test_store_not_cacheable_cachecontrol_no_cache(self):
        storage = DummyStorage()
        policy = self._makeOne(storage)
        environ = self._makeEnviron()
        headers = [ ('cache-control', 'no-cache') ]
        result = policy.store('200 OK', headers, environ)
        self.assertEqual(result, None)

    def test_store_not_cacheable_non_2XX_response(self):
        storage = DummyStorage()
        policy = self._makeOne(storage)
        environ = self._makeEnviron()
        result = policy.store('500 Error', [], environ)
        self.assertEqual(result, None)

    def test_store_allowed_request_method_cacheable(self):
        storage = DummyStorage()
        policy = self._makeOne(storage)
        policy.allowed_methods = ('FOO',)
        environ = self._makeEnviron()
        environ['REQUEST_METHOD'] = 'FOO'
        from email.Utils import formatdate
        now = formatdate()
        result = policy.store('200 OK', [('Date', now)], environ)
        self.assertEqual(result, None)
        self.assertEqual(storage.url, 'http://example.com')
        self.assertEqual(storage.status, '200 OK')
        self.assertEqual(storage.outheaders, [('Date', now)])

    def test_store_no_request_method_cacheable(self):
        storage = DummyStorage()
        policy = self._makeOne(storage)
        environ = self._makeEnviron()
        del environ['REQUEST_METHOD']
        from email.Utils import formatdate
        now = formatdate()
        result = policy.store('200 OK', [('Date', now)], environ)
        self.assertEqual(result, None)
        self.assertEqual(storage.url, 'http://example.com')
        self.assertEqual(storage.status, '200 OK')
        self.assertEqual(storage.outheaders, [('Date', now)])

    def test_store_get_request_method_cacheable(self):
        storage = DummyStorage()
        policy = self._makeOne(storage)
        environ = self._makeEnviron()
        from email.Utils import formatdate
        now = formatdate()
        result = policy.store('200 OK', [('Date', now)], environ)
        self.assertEqual(result, None)
        self.assertEqual(storage.url, 'http://example.com')
        self.assertEqual(storage.status, '200 OK')
        self.assertEqual(storage.outheaders, [('Date', now)])

    def test_fetch_fails_post_request_method(self):
        storage = DummyStorage(result=123)
        policy = self._makeOne(storage)
        environ = self._makeEnviron()
        environ['REQUEST_METHOD'] = 'POST'
        result = policy.fetch(environ)
        self.failIfEqual(result, 123)

    def test_fetch_fails_pragma_no_cache(self):
        storage = DummyStorage(result=123)
        policy = self._makeOne(storage)
        environ = self._makeEnviron()
        environ['HTTP_PRAGMA'] = 'no-cache'
        result = policy.fetch(environ)
        self.failIfEqual(result, 123)

    def test_fetch_fails_cachecontrol_no_cache(self):
        storage = DummyStorage(result=123)
        policy = self._makeOne(storage)
        environ = self._makeEnviron()
        environ['HTTP_PRAGMA'] = 'no-cache'
        result = policy.fetch(environ)
        self.failIfEqual(result, 123)

    def test_fetch_succeeds_no_request_method(self):
        from email.Utils import formatdate
        now = formatdate()
        cc = 'max-age=4000'
        expected = (200, [('Date', now), ('Cache-Control', cc)], [], [], [])
        storage = DummyStorage(result=[expected])
        policy = self._makeOne(storage)
        environ = self._makeEnviron()
        del environ['REQUEST_METHOD']
        result = policy.fetch(environ)
        self.assertEqual(result, expected[:3])

    def test_fetch_succeeds_get_request_method(self):
        from email.Utils import formatdate
        now = formatdate()
        cc = 'max-age=4000'
        expected = (200, [('Date', now), ('Cache-Control', cc)], [], [], [])
        storage = DummyStorage(result=[expected])
        policy = self._makeOne(storage)
        environ = self._makeEnviron()
        result = policy.fetch(environ)
        self.assertEqual(result, expected[:3])

class TestAcceleratorMiddleware(unittest.TestCase):
    def _getTargetClass(self):
        from repoze.accelerator.middleware import Accelerator
        return Accelerator

    def _makeOne(self, app, policy):
        klass = self._getTargetClass()
        return klass(app, policy)

    def _makeEnviron(self):
        return {}

    def test_call_fetch_from_cache(self):
        app = DummyApp()
        policy = DummyPolicy(result=('200 OK', [], ['abc', 'def']))
        environ = self._makeEnviron()
        accelerator = self._makeOne(app, policy)
        start_response = DummyStartResponse()
        generator = accelerator(environ, start_response)
        self.assertEqual(list(generator), ['abc', 'def'])
        self.assertEqual(start_response.status, '200 OK')
        self.assertEqual(start_response.headers,
                         [('X-Cached-By', 'repoze.accelerator')])
        self.assertEqual(start_response.exc_info, None)

    def test_call_nofetch_start_response_not_called(self):
        app = DummyApp()
        app.call_start_response = False
        policy = DummyPolicy(result=None)
        environ = self._makeEnviron()
        accelerator = self._makeOne(app, policy)
        start_response = DummyStartResponse()
        result = accelerator(environ, start_response)
        self.assertRaises(RuntimeError, list, result)

    def test_call_cantstore(self):
        app = DummyApp(headers=[('a', 'b')])
        policy = DummyPolicy(result=None)
        policy.handler = None
        environ = self._makeEnviron()
        accelerator = self._makeOne(app, policy)
        start_response = DummyStartResponse()
        result = list(accelerator(environ, start_response))
        self.assertEqual(result, ['hello', 'world'])
        self.assertEqual(start_response.status, '200 OK')
        self.assertEqual(start_response.headers, [('a', 'b')])
        self.assertEqual(start_response.exc_info, None)

    def test_call_canstore(self):
        app = DummyApp(headers=[('a', 'b')])
        policy = DummyPolicy(result=None)
        policy.handler = DummyHandler()
        environ = self._makeEnviron()
        accelerator = self._makeOne(app, policy)
        start_response = DummyStartResponse()
        result = list(accelerator(environ, start_response))
        self.assertEqual(result, ['hello', 'world'])
        self.assertEqual(start_response.status, '200 OK')
        self.assertEqual(start_response.headers, [('a', 'b')])
        self.assertEqual(start_response.exc_info, None)
        self.assertEqual(policy.handler.chunks, ['hello', 'world'])
        self.assertEqual(policy.handler.closed, True)


class Test_main(unittest.TestCase):

    def _callFUT(self, app, global_conf, **local_conf):
        from repoze.accelerator.middleware import main
        return main(app, global_conf, **local_conf)

    def _makeApp(self):
        return object()

    def test_main_defaults(self):
        from repoze.accelerator.middleware import NaivePolicy
        from repoze.accelerator.middleware import RAMStorage
        app = self._makeApp()

        accel = self._callFUT(app, {})

        self.failUnless(accel.app is app)
        self.failUnless(isinstance(accel.policy, NaivePolicy))
        self.failUnless(isinstance(accel.policy.storage, RAMStorage))

    def test_main_factories(self):

        app = self._makeApp()

        accel = self._callFUT(app,
                              {},
                              storage=_makeStorage,
                              policy=_makePolicy,
                             )

        self.failUnless(accel.app is app)
        self.failUnless(isinstance(accel.policy, _Policy))
        self.failUnless(isinstance(accel.policy.config, dict))
        self.failUnless(isinstance(accel.policy.storage.config, dict))

    def test_main_entry_points(self):

        app = self._makeApp()

        accel = self._callFUT(app,
                              {},
                              storage='repoze.accelerator.tests:_makeStorage',
                              policy='repoze.accelerator.tests:_makePolicy',
                             )

        self.failUnless(accel.app is app)
        self.failUnless(isinstance(accel.policy, _Policy))
        self.failUnless(isinstance(accel.policy.config, dict))
        self.failUnless(isinstance(accel.policy.storage.config, dict))

class _Storage:
    config = None

def _makeStorage(config=None):
    storage = _Storage()
    if config is not None:
        storage.config = config
    return storage

class _Policy:
    config = None

    def __init__(self, storage):
        self.storage = storage

def _makePolicy(storage, config=None):
    policy = _Policy(storage)
    if config is not None:
        policy.config = config
    return policy

class DummyHandler:
    def __init__(self):
        self.chunks = []
        self.closed = False
    def write(self, chunk):
        self.chunks.append(chunk)
    def close(self):
        self.closed = True

class DummyStartResponse:
    def __call__(self, status, headers, exc_info=None):
        self.status = status
        self.headers = headers
        self.exc_info = exc_info

class DummyApp:
    def __init__(self, status='200 OK', headers=()):
        self.status = status
        self.headers = headers
        self.call_start_response = True

    def __call__(self, environ, start_response):
        self.environ = environ
        if self.call_start_response:
            start_response(self.status, self.headers)
        return ['hello', 'world']

class DummyPolicy:
    def __init__(self, result):
        self.result = result
        self.handler = None

    def fetch(self, environ):
        return self.result

    def store(self, status, headers, environ):
        return self.handler

class DummyStorage:
    def __init__(self, result=None, writer=None):
        self.result = result
        self.writer = writer

    def store(self, url, status, outheaders, req_discrims, env_discrims):
        self.url = url
        self.status = status
        self.outheaders = outheaders
        self.req_discrims = req_discrims
        self.env_discrims = env_discrims
        return self.writer

    def fetch(self, url):
        return self.result

class DummyLock:
    def __init__(self):
        self.acquired = 0
        self.released = 0

    def acquire(self):
        self.acquired += 1

    def release(self):
        self.released += 1



repoze/accelerator/tests/test_logger.py
import unittest

class TestLoggerFactory(unittest.TestCase):
    def _getFUT(self):
        from repoze.accelerator.logger import make_logger
        return make_logger

    def test_factory_provides_ILoggerFactory(self):
        from zope.interface.verify import verifyObject
        from repoze.accelerator.interfaces import ILoggerFactory
        f = self._getFUT()
        verifyObject(ILoggerFactory, f)

    def test_default_config(self):
        f = self._getFUT()
        result = f({})
        self.assertEqual(result, None)

    def test_with_envvar(self):
        import os
        import logging
        import sys
        envvar = 'ACCELERATOR_LOG'
        try:
            os.environ[envvar] = '1'
            f = self._getFUT()
            logger = f({})
            self.failUnless(isinstance(logger, logging.Logger))
            self.assertEqual(logger.handlers[0].stream, sys.stdout)
        finally:
            del os.environ[envvar]
            
    def test_with_stdout_filename(self):
        import logging
        import sys
        f = self._getFUT()
        config = {'logger.filename':'stdout'}
        logger = f(config)
        self.failUnless(isinstance(logger, logging.Logger))
        self.assertEqual(logger.handlers[0].stream, sys.stdout)
        
    def test_with_stderr_filename(self):
        import logging
        import sys
        f = self._getFUT()
        config = {'logger.filename':'stderr'}
        logger = f(config)
        self.failUnless(isinstance(logger, logging.Logger))
        self.assertEqual(logger.handlers[0].stream, sys.stderr)

    def test_with_stderr_filename(self):
        import logging
        import sys
        f = self._getFUT()
        config = {'logger.filename':'stderr'}
        logger = f(config)
        self.failUnless(isinstance(logger, logging.Logger))
        self.assertEqual(logger.handlers[0].stream, sys.stderr)

    def test_with_real_filename(self):
        import os
        import tempfile
        tfn = tempfile.mktemp()
        try:
            import logging
            f = self._getFUT()
            config = {'logger.filename':tfn}
            logger = f(config)
            self.failUnless(isinstance(logger, logging.Logger))
            self.failUnless(isinstance(logger.handlers[0].stream, file))
        finally:
            os.remove(tfn)

    def test_override_level(self):
        import logging
        f = self._getFUT()
        config = {'logger.filename':'stderr',
                  'logger.log_level':'DEBUG'}
        logger = f(config)
        self.failUnless(isinstance(logger, logging.Logger))
        self.assertEqual(logger.level, logging.DEBUG)


repoze/accelerator/logger.py
from zope.interface import directlyProvides
from repoze.accelerator.interfaces import ILoggerFactory

def make_logger(config):
    import os
    import sys
    import logging

    if os.environ.get('ACCELERATOR_LOG'):
        log_stream = sys.stdout
        log_level = logging.DEBUG
        
    else:
        log_file = config.get('logger.filename', '')
        if not log_file or log_file.lower() == 'none':
            return None
        if log_file.lower() == 'stdout':
            log_stream = sys.stdout
        elif log_file.lower() == 'stderr':
            log_stream = sys.stderr
        else:
            log_stream = open(os.path.abspath(os.path.normpath(log_file)), 'a+')
        log_level = config.get('logger.log_level', 'INFO')
        log_level = log_level.upper()
        log_level = getattr(logging, log_level)

    handler = logging.StreamHandler(log_stream)
    fmt = '%(asctime)s %(message)s'
    formatter = logging.Formatter(fmt)
    handler.setFormatter(formatter)
    logger = logging.Logger('repoze.accelerator')
    logger.addHandler(handler)
    logger.setLevel(log_level)
    return logger
directlyProvides(make_logger, ILoggerFactory)


repoze/accelerator/policy.py
import calendar
from email.Utils import parsedate_tz
import time

from paste.request import construct_url
from paste.request import parse_headers
from paste.response import header_value

from zope.interface import implements
from zope.interface import directlyProvides

from repoze.accelerator.interfaces import IPolicy
from repoze.accelerator.interfaces import IPolicyFactory

class NullPolicy:
    """ Pass-through, caches nothing.
    """
    implements(IPolicy)

    def __init__(self):
        pass

    def fetch(self, environ):
        return None

    def store(self, status, headers, environ):
        pass

def make_null_policy(logger, storage, config):
    return NullPolicy()
directlyProvides(make_null_policy, IPolicyFactory)

class AcceleratorPolicy:
    """ Simple accelerating cache policy.

    - Allow configuration of "vary" policies for both request headers
      and request environment values.

    - Allow specification of "allowed methods"; requests which don't use
      one of these methods won't be served from cache.

    - Allow specification of "honor_shift_reload": if Pragma: no-cache or
      Cache-Control: no-cache exists in the request, and this value is true,
      the response will not be served from cache even if it otherwise might
      have been.

    - Allow specification of "store_https_responses".  If this is true,
      we will store https responses and some information provided by
      requests emitted via HTTPS.

    When deciding whether we can fetch from our storage or not:

    - If we honor shift-reload, and the request has a Pragma: no-cache
      or Cache-Control: no-cache associated with it, don't try to
      retrieve it from storage.

    - If the request method doesn't match one of our allowed_methods,
      don't try to retrieve it from storage.

    - If the request has a Range, header, don't try to retrieve it
      from storage.

    - If the request is conditional (e.g. If-Modified-Since,
      If-None-Match), don't try to retrieve it from storage.

    - Otherwise, attempt to retrieve it from storage.

    When deciding whether a value returned from storage should be
    checked for freshness:

    - If one representation stored matches the request headers and
      environment, check it for freshness.

    - Otherwise, abort.

    To decide whether an item is fresh or not:

    - If the entry in the cache is stale, don't serve from cache.
      Staleness is defined as having a CC: max-age < (now -
      entitydate) or an expires header whereby (expires - entitydate)
      < (now - entitydate).  Entities which don't have a Date header
      are also considered stale.

    When deciding whether we can store response data in our storage:

    - If the request method doesn't match one of our allowed_methods,
      don't store.

    - If the response status is not 200 (OK) or 203 (Non-Authoritative
      Information), don't store.

    - If the response has a Cache-Control header or a Pragma header,
      and either has 'no-cache' in its value, don't store.

    - If the response has a Cache-Control header, and it has a max-age
      of 0 or a max-age we don't understand, don't store.

    - If the request is an https request, and "store_https_responses" is false,
      don't store.

    - If the response does not have a Date header, assume the date is now.

    When storing data to storage:

    - Store the status, the end-to-end headers in the response, and
      information about request header and environment variance.
      
    """
    implements(IPolicy)

    def __init__(self,
                 logger,
                 storage,
                 allowed_methods=('GET',),
                 always_vary_on_headers=(),
                 always_vary_on_environ=('REQUEST_METHOD',),
                 honor_shift_reload=True,
                 store_https_responses=False,
                 ):
        self.logger = logger
        self.storage = storage
        self.allowed_methods = allowed_methods
        self.always_vary_on_headers = always_vary_on_headers
        self.always_vary_on_environ = always_vary_on_environ
        self.honor_shift_reload = honor_shift_reload
        self.store_https_responses = store_https_responses
        
    def fetch(self, environ):
        if environ.get('REQUEST_METHOD', 'GET') not in self.allowed_methods:
            return

        request_headers = list(parse_headers(environ))

        # if a Cache-Control/Pragma: no-cache header is in the request,
        # and if honor_shift_reload is true, we don't serve it from cache
        if self.honor_shift_reload:
            if self._check_no_cache(request_headers, environ):
                return
        # we don't try to serve range requests up from the cache
        if header_value(request_headers, 'Range'):
            return
        # we don't try to serve conditional requests up from cache
        for conditional in ('If-Modified-Since', 'If-None-Match',
                            'If-Match'):  # XXX other conditionals?
            if header_value(request_headers, conditional):
                return

        url = construct_url(environ)
        entries = self.storage.fetch(url)

        if entries:
            matching = self._discriminate(entries, request_headers, environ)
            if not matching:
                return

            now = time.time()
        
            discrims, expires, status, response_headers, body, extras = matching
            if expires > now:
                return status, response_headers, body

    def store(self, status, response_headers, environ):
        request_headers = list(parse_headers(environ))

        # abort if we shouldn't store this response
        request_method = environ.get('REQUEST_METHOD', 'GET')
        if request_method not in self.allowed_methods:
            return
        if not (status.startswith('200') or status.startswith('203')):
            return
        if environ['wsgi.url_scheme'] == 'https':
            if not self.store_https_responses:
                return
        if self._check_no_cache(response_headers, environ):
            return
        cc_header = header_value(response_headers, 'Cache-Control')
        if cc_header:
            cc_parts = parse_cache_control_header(cc_header)
            try:
                if int(cc_parts.get('max-age', '0')) == 0:
                    return
            except ValueError:
                return

        # if we didn't abort due to any condition above, store the response
        vary_header_names = []
        vary = header_value(response_headers, 'Vary')
        if vary is not None:
            vary_header_names.extend(
                [ x.strip().lower() for x in vary.split(',') ])
        if self.always_vary_on_headers:
            vary_header_names.extend(list(self.always_vary_on_headers))

        if '*' in vary_header_names:
            return

        discriminators = []
        for header_name in vary_header_names:
            value = header_value(request_headers, header_name)
            if value is not None:
                discriminators.append(('vary', (header_name, value)))
        for varname in self.always_vary_on_environ:
            value = environ.get(varname)
            if value is not None:
                discriminators.append(('env', (varname, value)))

        discriminators = tuple(sorted(discriminators))
        headers = endtoend(response_headers)
        url = construct_url(environ)

        # Response headers won't have a date if we aren't proxying to
        # another http server on our right hand side.
        date = header_value(response_headers, 'Date')
        if date is None:
            date = time.time()
        else:
            date = calendar.timegm(parsedate_tz(date))

        expires = self._expires(date, response_headers)

        # XXX purge?

        return self.storage.store(
            url,
            discriminators,
            expires,
            status,
            headers,
            )

    def _discriminate(self, entries, request_headers, environ):

        matching_entries = entries[:]

        for entry in entries:
            discrims, expires, status, headers, body, extras = entry
            for discrim in discrims:
                typ, (stored_name, stored_value) = discrim
                if typ == 'env':
                    strval = environ.get(stored_name)
                elif typ == 'vary':
                    strval = header_value(request_headers, stored_name)
                else:
                    raise ValueError(discrim)
                if strval is None or strval != stored_value:
                    matching_entries.remove(entry)
                    break

        if matching_entries:
            match = matching_entries[0] # this is essentially random
            return match

    def _check_no_cache(self, headers, environ):
        for nocache in ('Pragma', 'Cache-Control'):
            value = header_value(headers, nocache)
            if value and 'no-cache' in value.lower():
                return True
        return False

    def _expires(self, date, headers):
        cc_header = header_value(headers, 'Cache-Control')
        expires_header = header_value(headers, 'Expires')

        # logic stolen from httplib2
        if cc_header is not None:
            header_parts = parse_cache_control_header(cc_header)
            if 'max-age' in header_parts:
                try:
                    lifetime = int(header_parts['max-age'])
                    if lifetime == 0:
                        return date
                    return date + lifetime
                except ValueError:
                    return date

        if expires_header is not None:
            expires = parsedate_tz(expires_header)
            if expires is None:
                return date
            else:
                return calendar.timegm(expires)

    def _isfresh(self, date, headers):
        now = time.time()
        current_age = max(0, now - date)
        expires = self._expires(date, headers)
        lifetime = max(0, expires - date)
        return lifetime > current_age

def make_accelerator_policy(logger, storage, config):
    allowed_methods = config.get('policy.allowed_methods', 'GET')
    allowed_methods = [x.upper() for x in
                       filter(None, allowed_methods.split()) ]
    honor_shift_reload = config.get('policy.honor_shift_reload', False)
    honor_shift_reload = asbool(honor_shift_reload)
    store_https_responses = config.get('policy.store_https_responses',False)
    store_https_responses = asbool(store_https_responses)
    always_vary_on_headers = config.get('policy.always_vary_on_headers', '')
    always_vary_on_headers = filter(None, always_vary_on_headers.split())
    always_vary_on_environ = config.get('policy.always_vary_on_environ',
                                        'REQUEST_METHOD')
    always_vary_on_environ = filter(None, always_vary_on_environ.split())
    return AcceleratorPolicy(
        logger,
        storage,
        allowed_methods,
        always_vary_on_headers,
        always_vary_on_environ,
        honor_shift_reload,
        store_https_responses,
        )
directlyProvides(make_accelerator_policy, IPolicyFactory)

HOP_BY_HOP = ['connection',
              'keep-alive',
              'proxy-authenticate',
              'proxy-authorization',
              'te',
              'trailers',
              'transfer-encoding',
              'upgrade']

def endtoend(headers):
    connection_header = header_value(headers, 'Connection') or ''
    hop_by_hop = [x.strip().lower() for x in connection_header.split(',')]
    hop_by_hop.extend(HOP_BY_HOP)
    header_names = [ header[0] for header in headers ]
    return [(name, header_value(headers, name)) for name
             in header_names if name.lower() not in hop_by_hop]

def parse_cache_control_header(header):
    cc_parts = {}
    if header is not None:
        parts = [ x.strip() for x in header.split(',') ]
        for part in parts:
            if '=' in part:
                key, val = [ x.strip() for x in part.split('=', 1) ]
                cc_parts[key] = val
            else:
                cc_parts[part] = None
    return cc_parts

def asbool(val):
    val= str(val)
    if val.lower() in ('y', 'yes', 'true', 't'):
        return True
    return False

